### ISSUE TITLE
make navigation extendable

### ISSUE THREAD

**Issue by Toflar** [contao/core#3162]
Hi Leo

One thing that I really find annoying in Contao is that the navigation module is limited to the tl_page table and there's no way to hook upon.

Some scenarios:
- The catalog provides the possibility to generate a navigation out of a catalog but I can't tell the catalog to add its generated navigation *_into_ Home->Products so it in fact generates a submenu of "Products"
- Isotope could have the same use case as the catalog
- I wrote a customer specific module where the editors can describe sports. Every sport has to have its own navigation point and the two subpages "Description" and "Rules". The whole thing should be generated automatically. No problem at all except for adding the sports navigation to Home->Association->Sports.
- I could imagine that this could be very interesting when writing a front end admin panel for any software. Let's say you want to integrate any API from a third party application and provide the functionalities of this application within the default Contao navigation module. Same problem here...

I don't think that a hook in the current navigation module is the way to go because that would end up in too many useless hook calls (and you'd need to run through every registered hook for every page).
I rather think of a new library called "Navigation" that controls a navigation array (that can be extended - that's the point) and all the ModuleNavigation has to do would be

```
$this->Template->items = Navigation::generate($trail[$level]);
```

The fact that there exist several extensions (found "xNavigation" and "backboneit_navigation", maybe there are others) that try to solve this issue (but with hooks...) should also be evidence of the need to extend the navigation of Contao.

What do you think? ;)

--- _Originally created on June 14th, 2011, at 10:06am (ID 3162)_

**Comment by Toflar** [comment]
Came here because of the recent blog post. Checked the issue and I still think our navigation module should be improved.

**Comment by leofeyer** [comment]
Let's discuss this in our next call.

**Comment by leofeyer** [comment]
As discussed in Mumble on January 26th, @backbone87 is going to start a PR with a new navigation module for Contao 4.

**Comment by leofeyer** [comment]
This ticket appears to have stalled. Can it be closed or do you want to start/keep working on it?

**Comment by Toflar** [comment]
Same as ever: I still think it would make sense to improve the module but we should probably go for knp menu here, too. Maybe something for @sheeep or @backbone87 :)

**Comment by frontendschlampe** [comment]
It would be great if there will be more options for navigation. We already have an extended navigation module for Contao https://github.com/backbone87/contao-navigation ... I think there are some things which could be implemented ... @backbone87 should think about it!

**Comment by backbone87** [comment]
i agree with @Toflar -> we should somehow facilitate knp menu, since we already use it in the BE. the question is how we handle twig templates in the FE?

the codebase for my navigation is stable but somewhat dated. so a little bit of polishing isnt enough imho.

**Comment by leofeyer** [comment]
Contao 4 uses the KnpMenuBundle now, so I guess this ticket can be closed.

**Comment by fritzmg** [comment]
> Contao 4 uses the KnpMenuBundle now, so I guess this ticket can be closed.

Only for the back end, not for the front end. This would still be a cool feature to have for the front end navigation module.
